File: regions/system/library/go/messaging/noop.go
package messaging

import (
	"context"
	"sync"
)

// NoOpEventProducer はテスト用の no-op EventProducer 実装。
// 送信されたイベントを記録するが、実際の Kafka への送信は行わない。
// 並行アクセスに対して mutex で保護する。
type NoOpEventProducer struct {
	// mu は Published と closed への並行アクセスを保護する。
	mu sync.Mutex
	// Published は送信されたイベントのリスト（テスト検証用）。
	Published []EventEnvelope
	// Err は Publish 時に返すエラー（nil の場合はエラーなし）。
	Err error
	// closed はクローズ済みかどうか。
	closed bool
}

// Publish はイベントを記録する。
func (n *NoOpEventProducer) Publish(ctx context.Context, event EventEnvelope) error {
	n.mu.Lock()
	defer n.mu.Unlock()

	if n.Err != nil {
		return n.Err
	}
	n.Published = append(n.Published, event)
	return nil
}

// PublishBatch は複数イベントを順次 Publish する。
func (n *NoOpEventProducer) PublishBatch(ctx context.Context, events []EventEnvelope) error {
	for _, event := range events {
		if err := n.Publish(ctx, event); err != nil {
			return err
		}
	}
	return nil
}

// Close はプロデューサーをクローズ済みにする。
func (n *NoOpEventProducer) Close() error {
	n.mu.Lock()
	defer n.mu.Unlock()

	n.closed = true
	return nil
}

// IsClosed はプロデューサーがクローズ済みかどうかを返す。
func (n *NoOpEventProducer) IsClosed() bool {
	n.mu.Lock()
	defer n.mu.Unlock()

	return n.closed
}

// PublishedCount は送信されたイベント数を返す。
func (n *NoOpEventProducer) PublishedCount() int {
	n.mu.Lock()
	defer n.mu.Unlock()

	return len(n.Published)
}

File: regions/system/library/python/messaging/src/k1s0_messaging/noop.py
"""テスト用 NoOp プロデューサー"""

from __future__ import annotations

from .models import EventEnvelope
from .producer import EventProducer


class NoOpEventProducer(EventProducer):
    """テスト用の NoOp イベントプロデューサー。発行したメッセージを記録する。"""

    def __init__(self) -> None:
        self.published: list[EventEnvelope] = []
        self.closed = False

    def publish(self, envelope: EventEnvelope) -> None:
        self.published.append(envelope)

    async def publish_async(self, envelope: EventEnvelope) -> None:
        self.published.append(envelope)

    def publish_batch(self, envelopes: list[EventEnvelope]) -> None:
        self.published.extend(envelopes)

    def close(self) -> None:
        self.closed = True
